tests/integration-service/namespace-backed-environments.go

```diff
@@ -0,0 +1,311 @@
+
+			app, err := f.AsKubeAdmin.HasController.CreateApplication(applicationName, testNamespace)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(utils.WaitUntil(f.AsKubeAdmin.HasController.ApplicationGitopsRepoExists(app.Status.Devfile), 30*time.Second)).To(
+				Succeed(), fmt.Sprintf("timed out waiting for gitops content to be created for app %s in namespace %s: %+v", app.Name, app.Namespace, err),
+			)
+		}
+
+		createComponent := func() {
+			componentName = fmt.Sprintf("integration-suite-test-component-git-source-%s", util.GenerateRandomString(4))
+			timeout = time.Minute * 4
+			interval = time.Second * 1
+			// Create a component with Git Source URL being defined
+			// using cdq since git ref is not known
+			cdq, err := f.AsKubeAdmin.HasController.CreateComponentDetectionQuery(componentName, testNamespace, componentRepoURL, "", "", "", false)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(cdq.Status.ComponentDetected).To(HaveLen(1), "Expected length of the detected Components was not 1")
+
+			for _, compDetected := range cdq.Status.ComponentDetected {
+				originalComponent, err = f.AsKubeAdmin.HasController.CreateComponent(compDetected.ComponentStub, testNamespace, "", "", applicationName, true, map[string]string{})
+				Expect(err).NotTo(HaveOccurred())
+				Expect(originalComponent).NotTo(BeNil())
+				componentName = originalComponent.Name
+			}
+		}
+
+		cleanup := func() {
+			if !CurrentSpecReport().Failed() {
+				Expect(f.AsKubeAdmin.HasController.DeleteApplication(applicationName, testNamespace, false)).To(Succeed())
+				Expect(f.AsKubeAdmin.HasController.DeleteComponent(componentName, testNamespace, false)).To(Succeed())
+				integrationTestScenarios, err := f.AsKubeAdmin.IntegrationController.GetIntegrationTestScenarios(applicationName, testNamespace)
+				Expect(err).ShouldNot(HaveOccurred())
+
+				for _, testScenario := range *integrationTestScenarios {
+					Expect(f.AsKubeAdmin.IntegrationController.DeleteIntegrationTestScenario(&testScenario, testNamespace)).To(Succeed())
+				}
+				Expect(f.SandboxController.DeleteUserSignup(f.UserName)).To(BeTrue())
+			}
+		}
+
+		assertBuildPipelineRunFinished := func() {
+			timeout = time.Minute * 10
+			Eventually(func() error {
+				pipelineRun, err = f.AsKubeAdmin.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
+				if err != nil {
+					GinkgoWriter.Printf("PipelineRun has not been created yet for Component %s/%s\n", testNamespace, componentName)
+					return err
+				}
+				if !pipelineRun.HasStarted() {
+					return fmt.Errorf("pipelinerun %s/%s hasn't started yet", pipelineRun.GetNamespace(), pipelineRun.GetName())
+				}
+				return nil
+			}, timeout, constants.PipelineRunPollingInterval).Should(Succeed(), fmt.Sprintf("timed out when waiting for the PipelineRun to start for the component %s/%s", testNamespace, componentName))
```

Comment:
this logic is already included in the has controller method [WaitForComponentPipelineToBeFinished](https://github.com/redhat-appstudio/e2e-tests/blob/6f631c21e4de0cce2c2d74a3dcfec91ddaa265fc/pkg/utils/has/components.go#L88-L100) that is called on a line below

Reply:
Nice, then I should remove the `assertBuildPipelineRunFinished` function and replace all of its calls (within both the test files) with `WaitForComponentPipelineToBeFinished`.

Reply:
After testing this multiple times locally I realised that this block of code was doing something more than what  [WaitForComponentPipelineToBeFinished](https://github.com/redhat-appstudio/e2e-tests/blob/6f631c21e4de0cce2c2d74a3dcfec91ddaa265fc/pkg/utils/has/components.go#L88-L100) does. This `Eventually` block initializes the `pipelineRun` with the instance of build PLR as soon as it starts. This variable then gets used within the assertion 2 lines below the current Eventually block:
```
Expect(pipelineRun.Annotations["appstudio.openshift.io/snapshot"]).To(Equal(""))
```

Checking for the absence of the above annotation within the build PLR needs to be super quick right after the PLR finishes. And for that, we initialize the `pipelineRun` variable before the PLR finishes so that we won't need to make a `GetBuildPipelinerun()` call before the above assertion. For example:

--> This is flaky:
```
Expect(f.AsKubeDeveloper.HasController.WaitForComponentPipelineToBeFinished(originalComponent, "", 2, f.AsKubeAdmin.TektonController)).To(Succeed())
pipelineRun, _ = f.AsKubeAdmin.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
Expect(pipelineRun.Annotations["appstudio.openshift.io/snapshot"]).To(Equal(""))
```

--> This works:
```
pipelineRun, _ = f.AsKubeDeveloper.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
Expect(f.AsKubeDeveloper.HasController.WaitForComponentPipelineToBeFinished(originalComponent, "", 2, f.AsKubeAdmin.TektonController)).To(Succeed())
Expect(pipelineRun.Annotations["appstudio.openshift.io/snapshot"]).To(Equal(""))
```
* So, I've included the above approach within the tests, and removed this `Eventually` block.